test/general/openpgp.js

```diff
@@ -2588,6 +2703,118 @@ amnR6g==
+        multipleEncryptionAndSigningSubkeys
+      );
+
+      it('Encrypt message with a specific encryption key id', async function () {
+        const primaryKey = await getPrimaryKey();
+        let m;
+        let p;
+        for (let i = 0; i < encryptionKeyIds.length; i++) {
+          m = await openpgp.readArmoredMessage(await openpgp.encrypt({
+            message: openpgp.Message.fromText("Hello World\n"),
+            publicKeys: primaryKey,
+            encryptionKeyIds: [encryptionKeyIds[i]]
+          }));
+          p = m.packets.filterByTag(openpgp.enums.packet.publicKeyEncryptedSessionKey);
+          expect(p.length).equals(1);
+          expect(p[0].publicKeyId.equals(encryptionKeyIds[i])).to.be.true;
+        }
+      });
+
+      it('Sign message with a specific signing key id', async function () {
+        const primaryKey = await getPrimaryKey();
+        let s;
+        let p;
+        for (let i = 0; i < signingKeyIds.length; i++) {
+          s = await openpgp.readArmoredSignature(await openpgp.sign({
+            message: openpgp.Message.fromText("Hello World\n"),
+            privateKeys: primaryKey,
+            signingKeyIds: [signingKeyIds[i]],
+            detached: true
+          }));
+          p = s.packets.filterByTag(openpgp.enums.packet.signature);
+          expect(p.length).equals(1);
+          expect(p[0].issuerKeyId.equals(signingKeyIds[i])).to.be.true;
+        }
+      });
+
+      it('Encrypt and sign with specific encryption/signing key ids', async function () {
+        const primaryKey = await getPrimaryKey();
+        const latestEncryptionKeyId = (await primaryKey.getEncryptionKey()).getKeyId();
+        const latestSigningKeyId = (await primaryKey.getSigningKey()).getKeyId();
+        const plaintextMessage = openpgp.Message.fromText("Hello World\n");
+
+        const checkEncryptedPackets = (encryptionKeyIds, pKESKList) => {
+          pKESKList.forEach(({ publicKeyId }, i) => {
+            if (!encryptionKeyIds[i]) {
+              expect(publicKeyId.equals(latestEncryptionKeyId)).to.be.true;
+              return;
+            }
+            expect(publicKeyId.equals(encryptionKeyIds[i])).to.be.true;
+          });
+        };
+        const checkSignatures = (signingKeyIds, signatures) => {
+          signatures.forEach(({ keyid }, i) => {
+            if (!signingKeyIds[i]) {
+              expect(keyid.equals(latestSigningKeyId)).to.be.true;
+              return;
+            }
+            expect(keyid.equals(signingKeyIds[i])).to.be.true;
+          });
+        };
+
+        let d;
+        let m;
+        let pList;
+        let kIds;
+        let sIds;
+        // select every tuple of every 2 pair from each set.
+        for (let u = 0; u < encryptionKeyIds.length; u++) {
+          for (let v = 0; v < encryptionKeyIds.length; v++) {
+            for (let w = 0; w < signingKeyIds.length; w++) {
+              for (let z = 0; z < signingKeyIds.length; z++) {
```

Comment:
It looks like this test failed because it timed out, because it took longer than 10 seconds, probably because of the high number of combinations here. One way to fix this would be to try all the combinations in parallel (e.g. with `await Promise.all(...)`).

But, to be perfectly honest, I have the feeling that testing all the combinations probably doesn't really add much - testing the default is already done in other tests, for example, and I don't think there's a big chance that one combination would work and another wouldn't. The main thing that I feel this test adds is being able to encrypt (and sign) with multiple subkeys of the same key.

So I would personally reduce this test to that: pass the three encryption key IDs, and the three signing key IDs, and then afterwards, check that they match. That should simplify the test and make it a lot faster.

Reply:
I am often a fan of naive testing 😆 I went with this all 2-pair testing to ensure the correct order is always being executed, rather than defining a single test case that might pass because it was lucky enough to mask any logical error when it comes to selecting the corresponding signing/key id. Yes, I do believe this amount of tests are unnecessary, I will replace it with the above.